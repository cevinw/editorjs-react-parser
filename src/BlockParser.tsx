import React from "react";
import AlertBlock, { AlertConfig } from "./BlockParsers/alert";
import ParagraphBlock, { ParagraphConfig } from "./BlockParsers/paragraph";
import CodeBlock, { CodeConfig } from "./BlockParsers/code";
import DelimiterBlock, { DelimiterConfig } from "./BlockParsers/delimiter";
import HeaderBlock from "./BlockParsers/header";
import ImageBlock, { ImageConfig } from "./BlockParsers/image";
import ColumnsBlock, { ColumnsConfig } from "./BlockParsers/columns";
import QuoteBlock, { QuoteConfig } from "./BlockParsers/quote";
import TableBlock, { TableConfig } from "./BlockParsers/table";
import ListBlock, { ListConfig } from "./BlockParsers/list";
import ErrorBlock, { ErrorConfig } from "./BlockParsers/error";
import Raw from "./BlockParsers/raw";

export interface BlockParserProps {
  data: OutputData;
  config?: BlockParserConfig;
}

export interface BlockParserConfig {
  alert?: AlertConfig;
  code?: CodeConfig;
  columns?: ColumnsConfig;
  table?: TableConfig;
  image?: ImageConfig;
  paragraph?: ParagraphConfig;
  list?: ListConfig;
  quote?: QuoteConfig;
  delimiter?: DelimiterConfig;
  error?: ErrorConfig;
}

export interface OutputBlockData<Data extends object = any> {
  id?: BlockId;
  type: string;
  data: BlockToolData<Data>;
  tunes?: { [name: string]: BlockTuneData };
}

export interface OutputData {
  version?: string;
  time?: number;
  blocks: OutputBlockData[];
}

export type BlockId = string;
export type BlockToolData<T extends object = any> = T;
export type BlockTuneData = any;

export const BlockParser = ({ data, config }: BlockParserProps) =>
  data.blocks.map((blockData: OutputBlockData) => parseData(blockData, config));

const parseData = (
  item: OutputBlockData,
  config?: BlockParserConfig,
): React.JSX.Element => {
  return (
    {
      raw: <Raw key={item.id} item={item} />,
      header: <HeaderBlock key={item.id} item={item} />,
      code: <CodeBlock key={item.id} item={item} config={config?.code} />,
      list: <ListBlock key={item.id} item={item} config={config?.list} />,
      alert: <AlertBlock key={item.id} item={item} config={config?.alert} />,
      quote: <QuoteBlock key={item.id} item={item} config={config?.quote} />,
      table: <TableBlock key={item.id} item={item} config={config?.table} />,
      image: <ImageBlock key={item.id} item={item} config={config?.image} />,
      delimiter: (
        <DelimiterBlock key={item.id} item={item} config={config?.delimiter} />
      ),
      paragraph: (
        <ParagraphBlock key={item.id} item={item} config={config?.paragraph} />
      ),
      columns: (
        <ColumnsBlock
          key={item.id}
          item={item}
          config={config?.columns}
          blockRendererConfig={config}
        />
      ),
    }[item.type] || (
      <ErrorBlock key={item.id} item={item} config={config?.error} />
    )
  );
};

export default BlockParser;
