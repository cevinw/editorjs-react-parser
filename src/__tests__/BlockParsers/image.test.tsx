import { describe, expect } from "vitest";
import test from "../conftest";
import { OutputBlockData } from "../../BlockParser";
import { render } from "@testing-library/react";
import React from "react";
import ImageBlock from "../../BlockParsers/image";

describe("Image block tests", () => {
  test("Should parse image block", async () => {
    const inputItem: OutputBlockData = {
      id: "testid",
      type: "image",
      data: {
        file: {
          url: "http://test.test.png",
        },
        caption: "",
        withBorder: false,
        stretched: false,
        withBackground: false,
      },
    };
    const { container } = render(<ImageBlock item={inputItem} />);
    expect(container).toMatchSnapshot();
  });
});
