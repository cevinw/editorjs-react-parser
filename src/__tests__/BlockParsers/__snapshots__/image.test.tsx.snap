// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Image block tests > Should parse image block 1`] = `
<div>
  <section
    class="grid mt-2"
  >
    <img
      alt=""
      class="rounded justify-self-center dark:dark:opacity-75"
      src="http://test.test.png"
      style="width: 500px;"
    />
  </section>
</div>
`;
